test/Common/Test.hs

```diff
@@ -841,6 +841,46 @@ testSelectJoin run = do
               return p
           liftIO $ (entityVal <$> ps) `shouldBe` [p1]
 
+testSelectSubQuery :: Run -> Spec
+testSelectSubQuery run = do
+  describe "select subquery" $ do
+    it "works" $ do
+      run $ do
+        p1e <- insert' p1
+        let q = from $ \p -> do
+                  return ( p ^. PersonName, p ^. PersonAge)
+        ret <- select $ fromQuery q pure       
+        liftIO $ ret `shouldBe` [ (Value $ personName p1, Value $ personAge p1) ]
+
+    it "lets you order by alias" $ do
+      run $ do
+        _ <- insert' p1
+        _ <- insert' p3
+        let q = from $ \p -> do
+                  return ( p ^. PersonName, p ^. PersonAge)
+        ret <- select $ fromQuery q $ \(name, age) -> do 
+            orderBy [ asc age ]
+            pure name
+        liftIO $ ret `shouldBe` [ Value $ personName p3, Value $ personName p1 ]
+
+    it "supports groupBy" $ do
+      run $ do
+        l1k <- insert l1
+        l3k <- insert l3
+        mapM_ (\k -> insert $ Deed k l1k) (map show [1..3 :: Int])
+
+        mapM_ (\k -> insert $ Deed k l3k) (map show [4..10 :: Int])
+        let q = from $ \( lord `InnerJoin` deed ) -> do
+                on $ lord ^. LordId ==. deed ^. DeedOwnerId
+                return (lord ^. LordId, deed ^. DeedId)
+
+        (ret :: [(Value (Key Lord), Value Int)]) <- select $ fromQuery q $ \(lordId,deedId) -> do
+                 groupBy (lordId)
+                 return (lordId, count deedId)
+
+        liftIO $ ret `shouldMatchList` [ (Value l3k, Value 7)
+                                       , (Value l1k, Value 3) ]
```

Comment:
These test cases are great, but I think I'm a little confused on what this brings that is novel. Unless I'm mistaken, all of these can be written right now:

```haskell
-- groupBy
let 
  q = 
    from $ \(lord `InnerJoin` deed) -> do
    on $ lord ^. LordId ==. deed ^. DeedOwnerId
    return (lord ^. LordId, deed ^. DeedId)
select $ do
  (lordId, deedId) <- q
  groupBy lordId
  return (lordId, count deedId)
```

Can you write some tests or examples that demonstrate the utility that this approach brings?

Reply:
You are correct, all the examples in the tests currently are just proving that things that were possible before are possible with selection from a subquery, I will try to figure out a case that couldn't be trivially rewritten. However, this is mainly just foundational work to support union/subquery joining.

Reply:
That's cool! I think I want to see how union/subquery works out with this, or at least a path forward, before I merge it into the library. How much extra work do you imagine that might be?

Reply:
https://github.com/belevy/esqueleto/tree/monoidal-fromclause has my work towards joining on subqueries. It follows the syntax i used above trying to be minimally invasive to the way things work right now. It only supports a single join and that has to be the first thing you do. I am working towards making that work. 

My work for unions is at https://github.com/belevy/esqueleto/tree/union-prototype. That I feel is more complete and is useful right now.

Additionally I was able to come up with a test case that is at least not trivial without fromQuery though it may be possible I am not sure. It involves aggregating along two dimensions while limiting the result set after the first aggregation. "How many lords with more than 3 deeds?"

Reply:
fixed the single join only issue. But ordering still is incredibly important. All of the joins have to come before the on clauses start